exla/c_src/exla/exla.cc

```diff
@@ -511,6 +511,60 @@ ERL_NIF_TERM dynamic_update_slice(ErlNifEnv* env, int argc, const ERL_NIF_TERM a
   return exla::nif::ok(env, exla::nif::make<xla::XlaOp>(env, op));
 }
 
+ERL_NIF_TERM gather(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
+  if (argc != 7) {
+    return exla::nif::error(env, "Bad argument count.");
+  }
+
+  xla::XlaOp* operand;
+  xla::XlaOp* start_indices;
+  exla::int64 index_vector_dim;
+  std::vector<xla::int64> slice_sizes;
+  std::vector<xla::int64> offset_dims;
+  std::vector<xla::int64> collapsed_slice_dims;
+  std::vector<xla::int64> start_index_map;
+
+  if (!exla::nif::get<xla::XlaOp>(env, argv[0], operand)) {
+    return exla::nif::error(env, "Unable to get operand.");
+  }
+  if (!exla::nif::get<xla::XlaOp>(env, argv[1], start_indices)) {
+    return exla::nif::error(env, "Unable to get start indices.");
+  }
+  if (!exla::nif::get(env, argv[2], &index_vector_dim)) {
+    return exla::nif::error(env, "Unable to get index vector dim.");
+  }
+  if (!exla::nif::get_list(env, argv[3], slice_sizes)) {
+    return exla::nif::error(env, "Unable to get slice sizes.");
+  }
+  if (!exla::nif::get_list(env, argv[4], offset_dims)) {
+    return exla::nif::error(env, "Unable to get offset dims.");
+  }
+  if (!exla::nif::get_list(env, argv[5], collapsed_slice_dims)) {
+    return exla::nif::error(env, "Unable to get collapsed slice dims.");
+  }
+  if (!exla::nif::get_list(env, argv[6], start_index_map)) {
+    return exla::nif::error(env, "Unable to get start index map.");
+  }
+
+  xla::GatherDimensionNumbers dim_numbers;
+
+  dim_numbers.set_index_vector_dim(index_vector_dim);
+
+  for (auto dim : offset_dims) {
+    dim_numbers.add_offset_dims(dim);
+  }
+  for (auto dim : collapsed_slice_dims) {
+    dim_numbers.add_collapsed_slice_dims(dim);
+  }
+  for (auto dim : start_index_map) {
+    dim_numbers.add_start_index_map(dim);
+  }
```

Comment:
This is just a nit, but generally I've tried to move any XLA Proto object logic like Precision, ConvDimNos, etc. into `exla_nif_util.cc`.

Reply:
Ah I see! One concern I have, this consequently changes the API to `Op.gather(op, start_indices, dimension_nums, slice_sizes)` where `dimension_nums` is a tuple and it's not immediately clear how this tuple should look like. Thoughts?

Reply:
If it introduces unnecessary complexity then don't worry about making the change. I would just ship as is :D

Reply:
To give another example:

https://github.com/elixir-nx/nx/blob/c8353de695a3d70dc6d518e0ab2ec832faa7df68/exla/lib/exla/defn.ex#L306-L320

Here we pass a `{contract_axes1, batch_axes1, contract_axes2, batch_axes2}` tuple, and to figure out what the tuple should be, we probably need to look into `exla::nif::get_dot_dimension_numbers`.

If there's some benefit I'm missing then I can certainly make the change, otherwise feel free to merge =)